druid/src/widget/aspect_ratio_box.rs

```diff
@@ -0,0 +1,279 @@
+use druid::widget::prelude::*;
+use druid::Data;
+use tracing::{instrument, warn};
+
+/// A widget that preserves the aspect ratio given to it.
+///
+/// If given a child, this widget forces the child to have a width and height that preserves
+/// the aspect ratio.
+///
+/// If not given a child, The box will try to size itself  as large or small as possible
+/// to preserve the aspect ratio.
+pub struct AspectRatioBox<T> {
+    inner: Option<Box<dyn Widget<T>>>,
+    ratio: f64,
+}
+
+impl<T> AspectRatioBox<T> {
+    /// Create container with a child and aspect ratio.
+    ///
+    /// If aspect ratio <= 0.0, the ratio will be set to 1.0
+    pub fn new(inner: impl Widget<T> + 'static, ratio: f64) -> Self {
+        Self {
+            inner: Some(Box::new(inner)),
+            ratio: AspectRatioBox::<T>::clamp_ratio(ratio),
+        }
+    }
+
+    /// Create container without child but with an aspect ratio.
+    ///
+    /// If aspect ratio <= 0.0, the ratio will be set to 1.0
+    pub fn empty(ratio: f64) -> Self {
+        Self {
+            inner: None,
+            ratio: AspectRatioBox::<T>::clamp_ratio(ratio),
+        }
+    }
+
+    /// Set the ratio of the box.
+    ///
+    /// The ratio has to be a value between 0 and 1, excluding 0. It will be clamped
```

Comment:
this does not sound correct to me? An aspect ratio of 2.0 should mean that for your result, height is 2x width.

Reply:
I am assuming a value of 2.0 means width is 2x height. If it is 0.5 then height is 2x width. So i'm going aspect ratio is width / height. I think that's how it is assumed everywhere else.

Reply:
I think I want to stick to aspect ratio being width / height because it is used that way in most other imaging mediums as far as I'm aware.

Reply:
I'm just going by what we do with the [`BoxConstraints`](https://docs.rs/druid/0.7.0/druid/struct.BoxConstraints.html#method.constrain_aspect_ratio).

I don't have strong feelings here, but what's most important is that we are consistent and explicit; that means we should be talking very clearly in the docs here about what the argument represents.

cc @derekdreery, who did the `BoxConstraints` impl, and maybe has some thoughts.

Reply:
Ok yes definitely, I'll specifically say what I'm expecting as far as aspect ratio.

Reply:
After changing the implementation I think the function in `BoxConstraints` is trying to achieve something else.